akka-projection-grpc/src/main/scala/akka/projection/grpc/internal/EventPusher.scala

```diff
@@ -0,0 +1,244 @@
+/*
+ * Copyright (C) 2009-2023 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.projection.grpc.internal
+
+import akka.Done
+import akka.NotUsed
+import akka.actor.typed.ActorSystem
+import akka.annotation.InternalApi
+import akka.grpc.scaladsl.BytesEntry
+import akka.grpc.scaladsl.Metadata
+import akka.grpc.scaladsl.StreamResponseRequestBuilder
+import akka.grpc.scaladsl.StringEntry
+import akka.persistence.query.typed.EventEnvelope
+import akka.projection.ProjectionContext
+import akka.projection.grpc.internal.FilterStage.Filter
+import akka.projection.grpc.internal.ProtobufProtocolConversions.offsetToProtoOffset
+import akka.projection.grpc.internal.ProtobufProtocolConversions.updateFilterFromProto
+import akka.projection.grpc.internal.proto.ConsumeEventIn
+import akka.projection.grpc.internal.proto.ConsumeEventOut
+import akka.projection.grpc.internal.proto.ConsumerEventInit
+import akka.projection.grpc.internal.proto.EventConsumerServiceClient
+import akka.projection.grpc.internal.proto.FilteredEvent
+import akka.projection.grpc.internal.proto.KeepAlive
+import akka.projection.grpc.producer.scaladsl.EventProducer.EventProducerSource
+import akka.stream.Attributes
+import akka.stream.FlowShape
+import akka.stream.Inlet
+import akka.stream.Outlet
+import akka.stream.scaladsl.Flow
+import akka.stream.scaladsl.FlowWithContext
+import akka.stream.scaladsl.Sink
+import akka.stream.scaladsl.Source
+import akka.stream.stage.GraphStage
+import akka.stream.stage.GraphStageLogic
+import akka.stream.stage.InHandler
+import akka.stream.stage.OutHandler
+import org.slf4j.LoggerFactory
+
+import java.util
+import scala.collection.immutable
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.concurrent.Promise
+import scala.concurrent.duration.DurationInt
+
+/**
+ * INTERNAL API
+ *
+ * gRPC push protocol handler for the producing side
+ */
+private[akka] object EventPusher {
+
+  private val logger = LoggerFactory.getLogger(getClass)
+
+  def apply[Event](
+      originId: String,
+      client: EventConsumerServiceClient,
+      eps: EventProducerSource,
+      additionalRequestMetadata: Metadata)(implicit system: ActorSystem[_])
+      : FlowWithContext[EventEnvelope[Event], ProjectionContext, Done, ProjectionContext, NotUsed] = {
+    import akka.projection.grpc.internal.ProtobufProtocolConversions.transformAndEncodeEvent
+
+    val keepAliveTimeout = 5.seconds // FIXME config
+
+    implicit val ec: ExecutionContext = system.executionContext
+    val protoAnySerialization = new ProtoAnySerialization(system)
+    val topicFiltersPromise = Promise[immutable.Seq[proto.FilterCriteria]]()
```

Comment:
Dang, this isn't safe, we close over a promise created when creating the flow, but the flow may be materialized several times, each materialization should use the filter from its own connection, not a previous one.

Reply:
Fixed [59057fd](https://github.com/akka/akka-projection/pull/932/commits/59057fd1b9dba5dfd4f9c2535b36bf8f86477573)